src/sql/src/query_model/attribute/rejected_nulls.rs

```diff
@@ -277,3 +277,253 @@ fn case_and(expr: &BoxScalarExpr) -> Option<(&BoxScalarExpr, &BoxScalarExpr)> {
+                cref(input(r#box, 1), 1),
+                cref(input(r#box, 1), 2),
+            ]) // {rhs.#0, rhs.#1, rhs.#2}
+        });
+    }
+
+    #[test]
+    fn test_full_outer_join_1() {
+        let mut model = Model::new();
+        let g_id = add_get(&mut model);
+        let b_id = add_full_outer_join(&mut model, g_id, |lhs, rhs| {
+            vec![
+                // P0: (lhs.#0 >= rhs.#0)
+                exp::gte(exp::cref(lhs, 0), exp::cref(rhs, 0)),
+                // P1: (lhs.#1 <= rhs.#1)
+                exp::lte(exp::cref(lhs, 1), exp::cref(rhs, 1)),
+                // P2: (lhs.#2 != rhs.#2)
+                exp::not_eq(exp::cref(lhs, 2), exp::cref(rhs, 2)),
+            ]
+        });
+
+        assert_derived_attribute(&mut model, b_id, |_| {
+            HashSet::from([]) // {}
+        });
+    }
+
+    /// Adds a get box to the given model with a schema consisting of
+    /// for 32-bit nullable integers.
+    fn add_get(model: &mut Model) -> BoxId {
+        let get_id = model.make_box(qgm::get(0).into());
+
+        let mut b = model.get_mut_box(get_id);
+        b.add_column(exp::base(0, typ::int32(true)));
+        b.add_column(exp::base(1, typ::int32(true)));
+        b.add_column(exp::base(2, typ::int32(true)));
+        b.add_column(exp::base(3, typ::int32(true)));
+
+        get_id
+    }
+
+    /// Adds a select join to the model and attaches the given `predicates`.
+    /// The select box has a single input connected to the `src_id` box.
+    fn add_select<F>(model: &mut Model, src_id: BoxId, predicates: F) -> BoxId
+    where
+        F: FnOnce(QuantifierId) -> Vec<BoxScalarExpr>,
+    {
+        let tgt_id = model.make_box(Select::default().into());
+        let inp_id = model.make_quantifier(QuantifierType::Foreach, src_id, tgt_id);
+
+        if let BoxType::Select(ref mut b) = model.get_mut_box(tgt_id).box_type {
+            b.predicates.extend_from_slice(&predicates(inp_id));
+        }
+
+        tgt_id
+    }
+
+    /// Adds a full outer join to the model and attaches the given `predicates`.
+    /// Both inputs are connected to the `src_id` box.
+    fn add_full_outer_join<F>(model: &mut Model, src_id: BoxId, predicates: F) -> BoxId
+    where
+        F: FnOnce(QuantifierId, QuantifierId) -> Vec<BoxScalarExpr>,
+    {
+        let tgt_id = model.make_box(OuterJoin::default().into());
+        let lhs_id = model.make_quantifier(QuantifierType::PreservedForeach, src_id, tgt_id);
+        let rhs_id = model.make_quantifier(QuantifierType::PreservedForeach, src_id, tgt_id);
+
+        if let BoxType::OuterJoin(ref mut b) = model.get_mut_box(tgt_id).box_type {
+            b.predicates.extend_from_slice(&predicates(lhs_id, rhs_id));
+        }
+
+        tgt_id
+    }
+
+    /// Adds a left outer join to the model and attaches the given `predicates`.
+    /// Both inputs are connected to the `src_id` box.
+    fn add_left_outer_join<F>(model: &mut Model, src_id: BoxId, predicates: F) -> BoxId
```

Comment:
It really confused me that `add_left_outer_join` did not take a `lhs_box_id` and a `rhs_box_id`.

We should either:
1) leave a note that these `add_*` methods are specific to this set of tests and should never be copied or migrated into any general test utility.
2) rename the method to `add_left_outer_self_join`.

Reply:
I added the missing second argument.